### PR TITLE
Sqlserver

### PR DESCRIPTION
**Pull request by johan-v-r**

### CODE REVIEW

File: Butterfly.SqlServer/SqlServerDatabase.cs [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+					}
+				}
+
+				return rows.ToArray();
+			}, sql, executableParams);
+
+			return result;
+		}
+
+		private string ConvertLimit(string sql)
```
Comment (johan-v-r) [inline Butterfly.SqlServer/SqlServerDatabase.cs]: The need for this function, as well as the extra `executableSql += " SELECT SCOPE_IDENTITY();";` in `SqlServerTransaction.DoInsertAsync()`, should probably be removed once/if we override the `SelectStatement` and `InsertStatement`

**Comment by firesharkstudios** [comment]
Bah, didn't even realize LIMIT was MySQL specific.  I tend to think all the SELECT should be standard SQL and not specific to any database.  This might mean adding a maxRows like parameter to the Select() methods and having each implementation convert them appropriately.  Just thinking out loud a bit.

**Comment by firesharkstudios** [comment]
I see that sqlConnection.BeginTransaction() supports passing in an isolation level.

Does setting an isolation level on BeginTransaction() appropriately eliminate the need to create the transaction in SqlDatabase to execute the SELECTs?  My first reaction was it seemed odd to create a transaction in SqlDatabase.ExecuteCommandAsync() in order to execute SELECT statements.

I should probably spend time understanding how the MS SQL Server isolation levels work.

**Comment by johan-v-r** [comment]
> Bah, didn't even realize LIMIT was MySQL specific  

PostgreSQL also uses `LIMIT`, but then again Oracle uses `ROWNUM`  

___
> This might mean adding a maxRows like parameter to the Select() methods and having each implementation convert them appropriately.  

Yes 100% agree with you. Kinda what I meant with "_override the `SelectStatement` and `InsertStatement`_"

___
> Does setting an isolation level on BeginTransaction() appropriately eliminate the need to create the transaction in SqlDatabase to execute the SELECTs?  

Unfortunately not. Any new query requires an explicit transaction to override the default isolation level and not be blocked. There's also no way to set the database's default isolation level, and I don't think we should rely on that anyway.  
I don't think this is the perfect solution, but perhaps I'm missing something obvious.

**Comment by firesharkstudios** [comment]
So, running the test with LocalDB results in...

```
Butterfly.Core.Database.DatabaseException: Snapshot isolation transaction failed accessing database 'Butterfly' because snapshot isolation is not allowed in this database. Use ALTER DATABASE to allow snapshot isolation.
```

I thought LocalDB was SQL Server Express -- guess they turn off more features in LocalDB.

Do we have any other alternatives to using snapshot isolation on the SELECT queries?

**Comment by firesharkstudios** [comment]
So, I do get the tests to run successfully against LocalDB by...

1) Executing `ALTER DATABASE Butterfly SET ALLOW_SNAPSHOT_ISOLATION ON` once

2) Prepending `SET Transaction Isolation Level Snapshot` for all SELECT statements instead of using a transaction with the isolation level set

Honestly, #2 probably isn't necessary but I didn't retest with your original code.

This is all really just jumping through hoops to make the tests work.  Butterfly itself doesn't really depend on this behavior. 

I think the default behavior of MS SQL Server to hang SELECTs that depend on an uncompleted transactions is pretty bad.  In complex server code, it's not hard to imagine that happening.  A webserver using a threadpool to handle requests could easily exhaust the threadpool if each thread started hanging like this (not everything is converted to use async/await yet).  Guess I have a different opinion than the smart people that designed MS SQL Server.

**Pull request merged**